proposals/features-library.md

```diff
@@ -0,0 +1,76 @@
+
+Additionally, [members of the Feature authoring community have expressed interest in sharing code between Features](https://github.com/orgs/devcontainers/discussions/10).  Exploring the source code of self-published Features, it is evident that many authors have copy/pasted code from other Features.  This is a sign that there is a need for a more efficient way to share code between Features.
+
+# Goal
+
+Provide a generic and efficient way to share code between Features.
+
+# Proposal
+
+Add an `include` property to the `devcontainer-feature.json`.  This will be honored during the Feature [packaging stage](https://containers.dev/implementors/features-distribution/#packaging). The contents of the provided relative path is copied into the packaged Feature archive.
+
+Property | Type | Description
+--- | --- | ---
+`include` | `string[]` | An array of relative paths to package with the Feature. Paths are relative to root directory when packaging (often the git root directory).  If the element is a folder, it is copied recursively.
+
+## Example
+
+```json
+{
+    "name": "my-feature",
+    "version": "0.0.1",
+    "include": [
+        "utils/",
+    ]
+}
+```
+
+The preceeding example will recursively copy the contents of the the `utils` folder into the temporary folder used to package the Feature.  The `utils` folder will be included with the published Features.
+
+This functionality will enable sharing a common set of helper functions between Features.  The following example shows how a Feature could include a `utils` folder that contains a `common.sh` file.
+
+```bash
+# utils/common.sh
+function common_function() {
+   echo "Hello, '$1'"
+}
+```
+
+```bash
+# my-feature/install.sh
+
+# Include common functions
+source "./utils/common.sh"
+
+# Use common function
+common_function "devcontainers"
+```
+
+-----
+
+# Possible Extensions
+
+This section provides hints to possible future work that could extend the proposal above.  These are not part of the proposal, but are provided to help the community understand the potential future impact of this proposal.
+
+## Distributing shared library Features
+
+In an effort to reduce the need for authors to copy/paste auxillary scripts between Feature namespaces, provide the ability to publish "library" Features.  These Features would be published [following the Feature distribution spec](https://containers.dev/implementors/features-distribution/#distribution) under a new manifest type. 
+
+A "library" can be included in a Feature's `include` property as an alterative to a relative path. By themselves, a "library" Feature would not be installable.
```

Comment:
Similar, for this library feature, how would consumers locate these scripts on disk? Or would they be automatically sourced before execution of the consuming feature script?

Reply:
I think that's open for debate whether the user has the choice to source, or whether we try to auto-source whatever is included (might be trickier if we copy recursively, so perhaps we'd need to limit to include explicit script files)

Reply:
My example below has the consumer source explicitly at the top of their `install.sh`.

Reply:
I see that there are two options:

- No implicit sourcing.  The `include` simply copies the files into the archive.  It's up to the author to decide how and where to source/execute/copy those resources.
-  Treat script files (likely by looking for `.sh` and potentially other extensions) differently, and automatically source them in the feature-install.sh preamble script before executing the author's `install.sh`

Reply:
No implicit sourcing is more flexible (e.g., scripts might be written like commands that immediately do something). 👍